File: fsm/fsm_machine.c
//
//  fsm_machine.c
//  FiniteStateMachine
//
//  Created by Moky on 14-12-12.
//  Copyright (c) 2014 Slanissue.com. All rights reserved.
//

#include <stdlib.h>
#include <string.h>

#include "fsm_chain_table.h"
#include "fsm_state.h"
#include "fsm_machine.h"

fsm_machine * fsm_machine_create()
{
	fsm_machine * m = (fsm_machine *)malloc(sizeof(fsm_machine));
	memset(m, 0, sizeof(fsm_machine));
	m->states = fsm_chain_create();
	m->current = FSMNotFound;
	return m;
}

void fsm_machine_destroy(fsm_machine * m)
{
	// 1. destroy the chain table for states
	fsm_chain_destroy(m->states);
//	m->states = NULL;
//	m->enter = NULL;
//	m->exit = NULL;
//	m->object = NULL;
	
	// 2. free the machine
	free(m);
}

void fsm_machine_add_state(fsm_machine * m, const fsm_state * s)
{
	fsm_chain_add(m->states, s);
}

const fsm_state * fsm_machine_get_state(const fsm_machine * m, unsigned int index)
{
	fsm_chain_node * node = fsm_chain_at(m->states, index);
	return fsm_chain_get(node);
}

const fsm_state * fsm_machine_get_state_by_name(const fsm_machine * m, const char * name, unsigned int * index)
{
	const fsm_state * s = NULL;
	*index = FSMNotFound;
	if (name) {
		const fsm_chain_node * node;
		unsigned int i = 0;
		DS_FOR_EACH_CHAIN_NODE(m->states, node) {
			s = fsm_chain_get(node);
			if (strcmp(s->name, name) == 0) {
				*index = i;
				break;
			}
			++i;
			s = NULL;
		}
	}
	return s;
}

void fsm_machine_change_state(fsm_machine * m, const char * name)
{
	// get index for new state with name
	unsigned int index = FSMNotFound;
	const fsm_state * s = fsm_machine_get_state_by_name(m, name, &index);
	
	if (index == m->current) {
		// state not change
		return;
	}
	
	fsm_chain_node * node = fsm_chain_at(m->states, index);
	const fsm_state * o = fsm_chain_get(node);
	if (o && m->exit) m->exit(m, o); // exit the old state
	
	m->current = index;
	if (s && m->enter) m->enter(m, s); // enter the new state
}

void fsm_machine_tick(fsm_machine * m)
{
	fsm_chain_node * node = fsm_chain_at(m->states, m->current);
	const fsm_state * s = fsm_chain_get(node);
	if (m && s) {
		fsm_state_tick(m, s);
	}
}

File: fsm/FSMMachine.m
//
//  FSMMachine.m
//  FiniteStateMachine
//
//  Created by Moky on 14-12-13.
//  Copyright (c) 2014 Slanissue.com. All rights reserved.
//

#import "fsm_machine.h"
#import "fsm_state.h"

#import "FSMState.h"
#import "FSMMachine.h"

extern fsm_state * inner_state(const FSMState * state);

static void _on_enter(const fsm_machine * m, const fsm_state * s)
{
	FSMMachine * machine = m->object;
	FSMState * state = s->object;
	[machine.delegate machine:machine enterState:state]; // call 'enterState:' of delegate
	[state onEnter:machine]; // call 'onEnter:' of state
}

static void _on_exit(const fsm_machine * m, const fsm_state * s)
{
	FSMMachine * machine = m->object;
	FSMState * state = s->object;
	[machine.delegate machine:machine exitState:state]; // call 'exitState:' of delegate
	[state onExit:machine]; // call 'onExit:' of state
}

typedef NS_ENUM(NSUInteger, FSMStatus) {
	FSMMachineStatusStop,
	FSMMachineStatusRunning,
	FSMMachineStatusPaused,
};

@interface FSMMachine () {
	
	FSMStatus _status;
}

@property(nonatomic, readwrite) fsm_machine * innerMachine;
@property(nonatomic, retain) NSMutableArray * states;

@end

@implementation FSMMachine

@synthesize innerMachine = _innerMachine;

@synthesize defaultStateName = _defaultStateName;
@synthesize states = _states;

@synthesize delegate = _delegate;

- (void) dealloc
{
	[_defaultStateName release];
	[_states release];
	
	if (_innerMachine) {
		fsm_machine_destroy(_innerMachine);
	}
	
	[super dealloc];
}

+ (instancetype) allocWithZone:(struct _NSZone *)zone
{
	id object = [super allocWithZone:zone];
	fsm_machine * m = fsm_machine_create();
	if (m) {
		m->enter = _on_enter;
		m->exit = _on_exit;
		m->object = object;
	}
	[object setInnerMachine:m];
	return object;
}

- (instancetype) init
{
	return [self initWithDefaultStateName:@"default" capacity:8];
}

/* designated initializer */
- (instancetype) initWithDefaultStateName:(NSString *)name capacity:(NSUInteger)capacity
{
	self = [super init];
	if (self) {
		_status = FSMMachineStatusStop;
		
		self.defaultStateName = name;
		self.states = [NSMutableArray arrayWithCapacity:capacity];
		
		self.delegate = nil;
	}
	return self;
}

- (void) addState:(FSMState *)state
{
	NSAssert([state isKindOfClass:[FSMState class]], @"error state: %@", state);
	if (!state) {
		return;
	}
	
	fsm_machine_add_state(_innerMachine, inner_state(state));
	[_states addObject:state];
}

- (void) changeState:(NSString *)name
{
	fsm_machine_change_state(_innerMachine, [name UTF8String]);
}

- (FSMState *) currentState
{
	const fsm_state * s = fsm_machine_get_state(_innerMachine, _innerMachine->current);
	NSAssert(s, @"failed to get current state: %d", _innerMachine->current);
	FSMState * state = s->object;
	NSAssert([state isKindOfClass:[FSMState class]], @"memory error");
	return state;
}

#pragma mark -

- (void) tick
{
	@synchronized(self) {
		fsm_machine_tick(_innerMachine);
	}
}

- (void) start
{
	switch (_status) {
		case FSMMachineStatusStop: {
			[self changeState:_defaultStateName];
			_status = FSMMachineStatusRunning;
			break;
		}
			
		case FSMMachineStatusRunning: {
			// already running
			break;
		}
			
		case FSMMachineStatusPaused: {
			[self changeState:_defaultStateName];
			_status = FSMMachineStatusRunning;
			break;
		}
			
		default: {
			// unknown status
			break;
		}
	}
}

- (void) stop
{
	switch (_status) {
		case FSMMachineStatusStop: {
			// already stop
			break;
		}
			
		case FSMMachineStatusRunning: {
			_status = FSMMachineStatusStop;
			[self changeState:nil];
			break;
		}
			
		case FSMMachineStatusPaused: {
			_status = FSMMachineStatusStop;
			[self changeState:nil];
			break;
		}
			
		default: {
			// unknown status
			break;
		}
	}
}

- (void) pause
{
	switch (_status) {
		case FSMMachineStatusStop: {
			// error
			break;
		}
			
		case FSMMachineStatusRunning: {
			_status = FSMMachineStatusPaused;
			
			FSMState * state = [self currentState];
			if ([_delegate respondsToSelector:@selector(machine:pauseState:)]) {
				[_delegate machine:self pauseState:state];
			}
			[state onPause:self];
			break;
		}
			
		case FSMMachineStatusPaused: {
			// already paused
			break;
		}
			
		default: {
			// unknown status
			break;
		}
	}
}

- (void) resume
{
	switch (_status) {
		case FSMMachineStatusStop: {
			// error
			break;
		}
			
		case FSMMachineStatusRunning: {
			// not paused
			break;
		}
			
		case FSMMachineStatusPaused: {
			_status = FSMMachineStatusRunning;
			
			FSMState * state = [self currentState];
			if ([_delegate respondsToSelector:@selector(machine:resumeState:)]) {
				[_delegate machine:self resumeState:state];
			}
			[state onResume:self];
			break;
		}
			
		default: {
			// unknown status
			break;
		}
	}
}

@end
